src/contracts.rs

```diff
@@ -333,21 +337,16 @@ fn generate_record_contract(
     let fields = properties.iter().map(|(name, schema)| {
         // try to convert to a contract, otherwise convert the predicate version
         // to a contract
-        let contract = if let Ok(c) = Contract::try_from(schema) {
-            c
-        } else {
-            contract_from_predicate(Predicate::from(schema))
-        };
+        let contract = Contract::try_from(schema)
+            .unwrap_or_else(|()| contract_from_predicate(Predicate::from(schema)));
```

Comment:
This won't work I think. We're using an error type of `()` for the `TryFrom` implementations, since we actually just want something isomorphic to `Option<_>`. We'd need to either ignore the value in the `Err` case, i.e. `|_|`, or explicitly say that we're expecting `()` like we're doing here. I like the latter, since it forces us to reevaluate the semantics if we ever produce a meaningful error type.

Reply:
My bad, I initially thought this was from an `Option`. Of course taking a unit argument and no argument isn't the same, so the proposed change isn't even "well typed"...